### PR TITLE
Move label filtering into KafkaPrometheusReporter and out of the KafkaPrometheusCollector

### PR DESCRIPTION
**Pull request by OwenCorrigan76**
This PR is in response to the following issue: https://github.com/strimzi/metrics-reporter/issues/9
Moving the label filtering into the `metricChange` method means that filtering is done when the Reporter is initialised. It was previously done in the `collect` method.

### CODE REVIEW

File: src/main/java/io/strimzi/kafka/metrics/KafkaMetricsCollector.java [hunk 1]
```diff
@@ -31,19 +29,14 @@
 public class KafkaMetricsCollector implements MultiCollector {
 
     private static final Logger LOG = LoggerFactory.getLogger(KafkaMetricsCollector.class);
-
-    private final Map<MetricName, KafkaMetric> metrics;
-    private final PrometheusMetricsReporterConfig config;
+    private final Map<MetricName, MetricWrapper> metrics;
     @SuppressFBWarnings({"UWF_FIELD_NOT_INITIALIZED_IN_CONSTRUCTOR"}) // This field is initialized in the setPrefix method
     private String prefix;
```
[the remark this answers is not in the document]
Comment (OwenCorrigan76) [reply src/main/java/io/strimzi/kafka/metrics/KafkaMetricsCollector.java]: Removed from Collector

File: src/main/java/io/strimzi/kafka/metrics/KafkaPrometheusMetricsReporter.java [hunk 2]
```diff
@@ -50,35 +53,44 @@ public KafkaPrometheusMetricsReporter() {
...
             metricChange(metric);
         }
     }
 
-    @Override
     public void metricChange(KafkaMetric metric) {
-        kafkaMetricsCollector.addMetric(metric);
+        String prometheusName = MetricWrapper.prometheusName(collector.getPrefix(), metric.metricName());
+        if (!config.isAllowed(prometheusName)) {
+            System.out.println("Is allowed has failed for: " + prometheusName);
```
[the remark this answers is not in the document]
Comment (OwenCorrigan76) [reply src/main/java/io/strimzi/kafka/metrics/KafkaPrometheusMetricsReporter.java]: Sorry, missed that line. Removed now

File: src/main/java/io/strimzi/kafka/metrics/KafkaPrometheusMetricsReporter.java [hunk 3]
```diff
@@ -29,11 +29,14 @@
 public class KafkaPrometheusMetricsReporter implements MetricsReporter {
 
     private static final Logger LOG = LoggerFactory.getLogger(KafkaPrometheusMetricsReporter.class);
-
     private final PrometheusRegistry registry;
-    @SuppressFBWarnings({"UWF_FIELD_NOT_INITIALIZED_IN_CONSTRUCTOR"}) // This field is initialized in the configure method
-    private KafkaMetricsCollector kafkaMetricsCollector;
-    @SuppressFBWarnings({"UWF_FIELD_NOT_INITIALIZED_IN_CONSTRUCTOR"}) // This field is initialized in the configure method
+    @SuppressFBWarnings({"UWF_FIELD_NOT_INITIALIZED_IN_CONSTRUCTOR"})
```
[the remark this answers is not in the document]
Comment (OwenCorrigan76) [reply src/main/java/io/strimzi/kafka/metrics/KafkaPrometheusMetricsReporter.java]: Fixed

File: src/main/java/io/strimzi/kafka/metrics/KafkaPrometheusMetricsReporter.java [hunk 4]
```diff
@@ -50,35 +53,44 @@ public KafkaPrometheusMetricsReporter() {
...
 
-    @Override
     public void metricChange(KafkaMetric metric) {
-        kafkaMetricsCollector.addMetric(metric);
+        String prometheusName = MetricWrapper.prometheusName(collector.getPrefix(), metric.metricName());
+        if (!config.isAllowed(prometheusName)) {
+            System.out.println("Is allowed has failed for: " + prometheusName);
+            LOG.trace("Ignoring metric {} as it does not match the allowlist", prometheusName);
+        } else {
+            MetricWrapper metricWrapper = new MetricWrapper(prometheusName, metric, metric.metricName().name());
```
[the remark this answers is not in the document]
Comment (OwenCorrigan76) [reply src/main/java/io/strimzi/kafka/metrics/KafkaPrometheusMetricsReporter.java:96]: Comments removed from `MetricWrapper`

File: src/test/java/io/strimzi/kafka/metrics/KafkaMetricsCollectorTest.java [hunk 5]
```diff
@@ -131,6 +106,12 @@ private void assertGaugeSnapshot(MetricSnapshot snapshot, double expectedValue,
         assertEquals(expectedLabels, datapoint.getLabels());
     }
 
+    private MetricWrapper newMetric(MetricName metricName, double value) {
```
[the remark this answers is not in the document]
Comment (OwenCorrigan76) [reply src/test/java/io/strimzi/kafka/metrics/KafkaMetricsCollectorTest.java]: Merged two methods into `buildMetric()`

File: src/test/java/io/strimzi/kafka/metrics/KafkaMetricsCollectorTest.java [hunk 6]
```diff
@@ -47,58 +43,50 @@ public void setup() {
...
+        collector.addMetric(metricName, newMetric(metricName, 3.0));
         metrics = collector.collect();
         assertEquals(1, metrics.size());
 
         MetricSnapshot updatedSnapshot = metrics.get(0);
         assertGaugeSnapshot(updatedSnapshot, 3.0, labels);
 
         // Removing the metric
-        collector.removeMetric(buildMetric("name", "group", 4.0));
+        collector.removeMetric(buildMetric("name", "group", 1.0));
```
[the remark this answers is not in the document]
Comment (OwenCorrigan76) [reply src/test/java/io/strimzi/kafka/metrics/KafkaMetricsCollectorTest.java]: Refactored to use `metricName`

File: src/test/java/io/strimzi/kafka/metrics/KafkaMetricsCollectorTest.java [hunk 7]
```diff
@@ -47,58 +43,50 @@ public void setup() {
...
     }
 
     @Test
     public void testCollectNonNumericMetric() {
-        Map<String, String> props = new HashMap<>();
-        props.put(PrometheusMetricsReporterConfig.ALLOWLIST_CONFIG, "kafka_server_group_name.*");
-        PrometheusMetricsReporterConfig config = new PrometheusMetricsReporterConfig(props, new PrometheusRegistry());
-        KafkaMetricsCollector collector = new KafkaMetricsCollector(config);
+        KafkaMetricsCollector collector = new KafkaMetricsCollector();
         collector.setPrefix("kafka.server");
```
[the remark this answers is not in the document]
Comment (OwenCorrigan76) [reply src/test/java/io/strimzi/kafka/metrics/KafkaMetricsCollectorTest.java]: Removed

File: src/test/java/io/strimzi/kafka/metrics/KafkaMetricsCollectorTest.java [hunk 8]
```diff
@@ -131,6 +106,12 @@ private void assertGaugeSnapshot(MetricSnapshot snapshot, double expectedValue,
         assertEquals(expectedLabels, datapoint.getLabels());
     }
 
+    private MetricWrapper newMetric(MetricName metricName, double value) {
+        KafkaMetric metric = buildMetric(metricName.name(), metricName.group(), value);
+        String prometheusName = MetricWrapper.prometheusName("kafka.server", metricName);
```
[the remark this answers is not in the document]
Comment (OwenCorrigan76) [reply src/test/java/io/strimzi/kafka/metrics/KafkaMetricsCollectorTest.java]: Extracted out to `METRIC_PREFIX`

File: src/test/java/io/strimzi/kafka/metrics/KafkaPrometheusMetricsReporterTest.java [hunk 9]
```diff
@@ -38,12 +38,14 @@ public void testLifeCycle() throws Exception {
         KafkaPrometheusMetricsReporter reporter = new KafkaPrometheusMetricsReporter(new PrometheusRegistry());
         Map<String, String> configs = new HashMap<>();
         configs.put(PrometheusMetricsReporterConfig.LISTENER_CONFIG, "http://:0");
+        configs.put(PrometheusMetricsReporterConfig.ALLOWLIST_CONFIG, "kafka_server_group_name.*");
```
[the remark this answers is not in the document]
Comment (OwenCorrigan76) [reply src/test/java/io/strimzi/kafka/metrics/KafkaPrometheusMetricsReporterTest.java:41]: Cheks added to see if allowList is working

File: src/test/java/io/strimzi/kafka/metrics/KafkaMetricsCollectorTest.java [hunk 10]
```diff
@@ -109,19 +97,6 @@ public void testCollectNonNumericMetric() {
         assertEquals(expectedLabels, snapshot.getDataPoints().get(0).getLabels());
     }
 
-    @Test
```
[the remark this answers is not in the document]
Comment (OwenCorrigan76) [reply src/test/java/io/strimzi/kafka/metrics/KafkaMetricsCollectorTest.java:112]: Moved to `MetricWrapper` and accessed from there

File: src/main/java/io/strimzi/kafka/metrics/KafkaPrometheusMetricsReporter.java [hunk 11]
```diff
@@ -48,10 +52,28 @@ public KafkaPrometheusMetricsReporter() {
...
+    }
+
+    /**
+     * This method is used to get the prefix that is used for metric names.
+     *
+     * @return The prefix used for metric names.
+     */
+    public String getPrefix() {
+        return prefix;
+    }
```
Comment (mimaison) [inline src/main/java/io/strimzi/kafka/metrics/KafkaPrometheusMetricsReporter.java:71]: We can remove these 2 methods and just set the `prefix` field in `contextChange()`.

File: src/test/java/io/strimzi/kafka/metrics/KafkaMetricsCollectorTest.java [hunk 12]
```diff
@@ -131,24 +107,28 @@ private void assertGaugeSnapshot(MetricSnapshot snapshot, double expectedValue,
         assertEquals(expectedLabels, datapoint.getLabels());
     }
 
-    private KafkaMetric buildMetric(String name, String group, double value) {
+    private MetricWrapper buildMetric(MetricName metricName, double value) {
```
Comment (mimaison) [inline src/test/java/io/strimzi/kafka/metrics/KafkaMetricsCollectorTest.java:110]: Should we call that method `newMetric()` so it matches `newNonNumericMetric()` and also the names used in `YammerMetricsCollectorTest`?

File: src/test/java/io/strimzi/kafka/metrics/KafkaMetricsCollectorTest.java [hunk 13]
```diff
@@ -20,19 +18,19 @@
...
 import static org.junit.jupiter.api.Assertions.assertEquals;
 import static org.junit.jupiter.api.Assertions.assertInstanceOf;
 
 public class KafkaMetricsCollectorTest {
-
     private final MetricConfig metricConfig = new MetricConfig();
     private final Time time = Time.SYSTEM;
     private Map<String, String> tagsMap;
     private Labels labels;
+    private static final String METRIC_PREFIX = "kafka.server";
```
Comment (mimaison) [inline src/test/java/io/strimzi/kafka/metrics/KafkaMetricsCollectorTest.java:32]: We typically put the `static final` fields above all other fields.

File: src/test/java/io/strimzi/kafka/metrics/KafkaMetricsCollectorTest.java [hunk 14]
```diff
@@ -20,19 +18,19 @@
...
+
```
Comment (mimaison) [inline src/test/java/io/strimzi/kafka/metrics/KafkaMetricsCollectorTest.java:33]: Let's remove this extra empty line.

File: src/test/java/io/strimzi/kafka/metrics/KafkaMetricsCollectorTest.java [hunk 15]
```diff
@@ -20,19 +18,19 @@
...
 
-import java.util.HashMap;
 import java.util.LinkedHashMap;
 import java.util.Map;
 
 import static org.junit.jupiter.api.Assertions.assertEquals;
 import static org.junit.jupiter.api.Assertions.assertInstanceOf;
 
 public class KafkaMetricsCollectorTest {
-
```
Comment (mimaison) [inline src/test/java/io/strimzi/kafka/metrics/KafkaMetricsCollectorTest.java:31]: Can we undo this change?

File: src/test/java/io/strimzi/kafka/metrics/KafkaPrometheusMetricsReporterTest.java [hunk 16]
```diff
@@ -38,31 +38,36 @@ public void testLifeCycle() throws Exception {
...
         reporter.configure(configs);
         reporter.contextChange(new KafkaMetricsContext("kafka.server"));
+
         Optional<Integer> port = reporter.getPort();
         assertTrue(port.isPresent());
         int initialMetrics = getMetrics(port.get()).size();
```
Comment (mimaison) [inline src/test/java/io/strimzi/kafka/metrics/KafkaPrometheusMetricsReporterTest.java:47]: Since each test now uses its own registry, we should be able to get rid of `initialMetrics` and always expect an empty registry initially.

File: src/test/java/io/strimzi/kafka/metrics/KafkaPrometheusMetricsReporterTest.java [hunk 17]
```diff
@@ -38,31 +38,36 @@ public void testLifeCycle() throws Exception {
...
+        // Adding a metric not matching the allowlist does nothing
+        KafkaMetric metric1 = buildMetric("other", "group", 0);
         reporter.init(Collections.singletonList(metric1));
-
         List<String> metrics = getMetrics(port.get());
-        assertEquals(initialMetrics + 1, metrics.size());
+        assertEquals(initialMetrics, metrics.size());
 
-        KafkaMetric metric2 = buildMetric("name2", "group", 0);
+        // Adding a metric not matching the allowlist does nothing
```
Comment (mimaison) [inline src/test/java/io/strimzi/kafka/metrics/KafkaPrometheusMetricsReporterTest.java:55]: This comment does not look right, this metric does match the allowlist

File: src/test/java/io/strimzi/kafka/metrics/KafkaPrometheusMetricsReporterTest.java [hunk 18]
```diff
@@ -38,31 +38,36 @@ public void testLifeCycle() throws Exception {
         KafkaPrometheusMetricsReporter reporter = new KafkaPrometheusMetricsReporter(new PrometheusRegistry());
         Map<String, String> configs = new HashMap<>();
         configs.put(PrometheusMetricsReporterConfig.LISTENER_CONFIG, "http://:0");
+        configs.put(PrometheusMetricsReporterConfig.ALLOWLIST_CONFIG, "kafka_server_group_name.*");
         reporter.configure(configs);
         reporter.contextChange(new KafkaMetricsContext("kafka.server"));
+
         Optional<Integer> port = reporter.getPort();
```
Comment (mimaison) [inline src/test/java/io/strimzi/kafka/metrics/KafkaPrometheusMetricsReporterTest.java:45]: Since we're touching this method, we can do `int port = reporter.getPort().orElseThrow();` to simplify the logic a bit.

File: src/test/java/io/strimzi/kafka/metrics/MetricWrapperTest.java [hunk 19]
```diff
@@ -10,28 +10,34 @@
...
+
+        Labels labels = MetricWrapper.labelsFromScope("k-1.v1.k_1.v2", "name");
+        assertEquals("k_1", PrometheusNaming.sanitizeLabelName("k-1"));
+        assertEquals("v1", labels.get("k_1"));
+        assertEquals(1, labels.size());
+    }
+
+    @Test
+    public void testLabelsFromTags() {
+        Map<String, String> tags = new LinkedHashMap<>();
```
Comment (mimaison) [inline src/test/java/io/strimzi/kafka/metrics/MetricWrapperTest.java:36]: Can we also try a regular scenario with multiple labels and no colliding names, like we have i https://github.com/strimzi/metrics-reporter/pull/38/files#diff-dc099f5ba5c7d3493db55fc8190b8badc3038f467d8edb8b97c3ca1022e3d319R22?

**Review by mimaison — COMMENTED** [review]
Thanks for the updates! I looks good overall. I left a few more suggestions but we're almost there.

File: src/main/java/io/strimzi/kafka/metrics/KafkaPrometheusMetricsReporter.java (hunk 11 above)
Comment (OwenCorrigan76) [reply src/main/java/io/strimzi/kafka/metrics/KafkaPrometheusMetricsReporter.java]: Removed

File: src/test/java/io/strimzi/kafka/metrics/KafkaMetricsCollectorTest.java (hunk 12 above)
Comment (OwenCorrigan76) [reply src/test/java/io/strimzi/kafka/metrics/KafkaMetricsCollectorTest.java]: Yes, changed to `newMetric()`

File: src/test/java/io/strimzi/kafka/metrics/KafkaMetricsCollectorTest.java (hunk 13 above)
Comment (OwenCorrigan76) [reply src/test/java/io/strimzi/kafka/metrics/KafkaMetricsCollectorTest.java]: Noted and moved

File: src/test/java/io/strimzi/kafka/metrics/KafkaMetricsCollectorTest.java (hunk 14 above)
Comment (OwenCorrigan76) [reply src/test/java/io/strimzi/kafka/metrics/KafkaMetricsCollectorTest.java]: Removed

File: src/test/java/io/strimzi/kafka/metrics/KafkaMetricsCollectorTest.java (hunk 15 above)
Comment (OwenCorrigan76) [reply src/test/java/io/strimzi/kafka/metrics/KafkaMetricsCollectorTest.java:31]: Undone

File: src/test/java/io/strimzi/kafka/metrics/KafkaPrometheusMetricsReporterTest.java (hunk 16 above)
Comment (OwenCorrigan76) [reply src/test/java/io/strimzi/kafka/metrics/KafkaPrometheusMetricsReporterTest.java]: Keeping initialMetrics due to JVM metrics being populated

File: src/test/java/io/strimzi/kafka/metrics/KafkaPrometheusMetricsReporterTest.java (hunk 17 above)
Comment (OwenCorrigan76) [reply src/test/java/io/strimzi/kafka/metrics/KafkaPrometheusMetricsReporterTest.java]: Fixed

File: src/test/java/io/strimzi/kafka/metrics/MetricWrapperTest.java (hunk 19 above)
Comment (OwenCorrigan76) [reply src/test/java/io/strimzi/kafka/metrics/MetricWrapperTest.java:36]: Chaged to test for the golden path then reset the map to check the behavior when names collide

**Review by mimaison — COMMENTED** [review]
I made another pass and left a few more minor suggestions

File: src/test/java/io/strimzi/kafka/metrics/KafkaMetricsCollectorTest.java [hunk 20]
```diff
@@ -131,24 +107,28 @@ private void assertGaugeSnapshot(MetricSnapshot snapshot, double expectedValue,
         assertEquals(expectedLabels, datapoint.getLabels());
     }
 
-    private KafkaMetric buildMetric(String name, String group, double value) {
+    private MetricWrapper newMetric(MetricName metricName, double value) {
         Measurable measurable = (config, now) -> value;
-        return new KafkaMetric(
+        KafkaMetric metric = new KafkaMetric(
```
[the remark this answers is not in the document]
Comment (OwenCorrigan76) [reply src/test/java/io/strimzi/kafka/metrics/KafkaMetricsCollectorTest.java]: Yes. Changed to `kafkaMetric`

File: src/test/java/io/strimzi/kafka/metrics/KafkaMetricsCollectorTest.java [hunk 21]
```diff
@@ -131,24 +107,28 @@ private void assertGaugeSnapshot(MetricSnapshot snapshot, double expectedValue,
...
                 new Object(),
-                new MetricName(name, group, "", tagsMap),
+                new MetricName(metricName.name(), metricName.group(), "", tagsMap),
```
[the remark this answers is not in the document]
Comment (OwenCorrigan76) [reply src/test/java/io/strimzi/kafka/metrics/KafkaMetricsCollectorTest.java]: Ah, I missed that. Now using `metricName`

File: src/test/java/io/strimzi/kafka/metrics/KafkaMetricsCollectorTest.java [hunk 22]
```diff
@@ -47,58 +45,49 @@ public void setup() {
...
+        KafkaMetricsCollector collector = new KafkaMetricsCollector();
 
         MetricSnapshots metrics = collector.collect();
         assertEquals(0, metrics.size());
 
-        // Adding a metric not matching the allowlist does nothing
-        collector.addMetric(buildMetric("name", "other", 2.0));
-        metrics = collector.collect();
-        assertEquals(0, metrics.size());
+        // Add a metric
```
[the remark this answers is not in the document]
Comment (OwenCorrigan76) [reply src/test/java/io/strimzi/kafka/metrics/KafkaMetricsCollectorTest.java]: Changed to `Adding` etc.

File: src/test/java/io/strimzi/kafka/metrics/KafkaPrometheusMetricsReporterTest.java [hunk 23]
```diff
@@ -38,39 +38,45 @@ public void testLifeCycle() throws Exception {
...
+        assertEquals(initialMetrics + 1, metrics.size());
+
         reporter.close();
     }
 
     @Test
     public void testMultipleReporters() throws Exception {
         Map<String, String> configs = new HashMap<>();
         configs.put(PrometheusMetricsReporterConfig.LISTENER_CONFIG, "http://:0");
+        configs.put(PrometheusMetricsReporterConfig.ALLOWLIST_CONFIG, "kafka_server_group_name.*");
```
[the remark this answers is not in the document]
Comment (OwenCorrigan76) [reply src/test/java/io/strimzi/kafka/metrics/KafkaPrometheusMetricsReporterTest.java]: Not needed here. Removed

File: src/main/java/io/strimzi/kafka/metrics/KafkaPrometheusMetricsReporter.java [hunk 24]
```diff
@@ -60,25 +64,30 @@ public void configure(Map<String, ?> map) {
...
+        registry.register(collector);
         for (KafkaMetric metric : metrics) {
             metricChange(metric);
         }
     }
 
-    @Override
     public void metricChange(KafkaMetric metric) {
-        kafkaMetricsCollector.addMetric(metric);
+        String prometheusName = MetricWrapper.prometheusName(this.prefix, metric.metricName());
```
[the remark this answers is not in the document]
Comment (OwenCorrigan76) [reply src/main/java/io/strimzi/kafka/metrics/KafkaPrometheusMetricsReporter.java]: Fixed

**Comment by OwenCorrigan76** [comment]
@mimaison Thanks for the extra pass and comments. Hopefully there's nothing missed now.

File: src/main/java/io/strimzi/kafka/metrics/KafkaPrometheusMetricsReporter.java [hunk 25]
```diff
@@ -29,12 +30,15 @@
 public class KafkaPrometheusMetricsReporter implements MetricsReporter {
 
     private static final Logger LOG = LoggerFactory.getLogger(KafkaPrometheusMetricsReporter.class);
-
     private final PrometheusRegistry registry;
     @SuppressFBWarnings({"UWF_FIELD_NOT_INITIALIZED_IN_CONSTRUCTOR"}) // This field is initialized in the configure method
-    private KafkaMetricsCollector kafkaMetricsCollector;
+    private KafkaMetricsCollector collector;
+    @SuppressFBWarnings({"UWF_FIELD_NOT_INITIALIZED_IN_CONSTRUCTOR"}) // This field is initialized in the init method
```
Comment (mimaison) [inline src/main/java/io/strimzi/kafka/metrics/KafkaPrometheusMetricsReporter.java:36]: `init` -> `configure`

File: src/main/java/io/strimzi/kafka/metrics/KafkaPrometheusMetricsReporter.java [hunk 26]
```diff
@@ -29,12 +30,15 @@
...
+    private PrometheusMetricsReporterConfig config;
     @SuppressFBWarnings({"UWF_FIELD_NOT_INITIALIZED_IN_CONSTRUCTOR"}) // This field is initialized in the configure method
     private Optional<HTTPServer> httpServer;
+    @SuppressFBWarnings({"UWF_FIELD_NOT_INITIALIZED_IN_CONSTRUCTOR"}) // This field is initialized in the setPrefix method
```
Comment (mimaison) [inline src/main/java/io/strimzi/kafka/metrics/KafkaPrometheusMetricsReporter.java:40]: `setPrefix` -> `contextChange`

**Review by scholzj — APPROVED** [review]
LGTM. Thanks.

**Review by mimaison — APPROVED** [review]
LGTM
